sklearn/ensemble/_forest.py

```diff
@@ -172,6 +172,12 @@ def _parallel_build_trees(
         indices = _generate_sample_indices(
             tree.random_state, n_samples, n_samples_bootstrap
         )
+
+        if indices.size == 0:
+            raise ValueError(
+                "No samples were selected from your data, try increasing max_samples."
```

Comment:
This is backward incompatible. The following works on `main`, but fails on this PR:

```python
from sklearn.datasets import load_iris
from sklearn.ensemble import RandomForestClassifier

X, y = load_iris(return_X_y=True)
rf = RandomForestClassifier(max_samples=1e-4)

rf.fit(X, y)
```

I can see that the trees are not interesting because they are nodes with no samples in them, but the `fit` call itself does not fail.

If we want the above to still work and want a better error for https://github.com/scikit-learn/scikit-learn/issues/24037, we can check `indices` only when `class_weight == "balanced_subsample"` a few lines below:

```python
elif class_weight == "balanced_subsample":
    if indices.size == 0:
        raise ValueError(...)
```

Reply:
Indeed, the exception is raised inside `compute_sample_weight()`, which doesn't get called with the default `class_weight` setting. So maybe we should handle the problem in `compute_sample_weight()`?

It would be easier to leave the case in your example code unchanged (`fit()` succeeds). I'm torn though if it isn't a bug/misleading users to think "yay it worked, I've got a classifier!" when in reality the RF they have is a weird thing with random(?) behaviour?

Maybe to keep the scope of this PR manageable: move the test into `compute_sample_weight()` and deal with the more philosophical question of "is this even a useful classifier?" in a new issue?

Reply:
> Maybe to keep the scope of this PR manageable: move the test into compute_sample_weight()

I like the idea of checking `indices` in `compute_sample_weight` itself. 

@anandhakrishnanaji Are you interested in updating this PR and move the check into `compute_sample_weight`?

Reply:
Looking at this again, I am okay with the breaking change described in https://github.com/scikit-learn/scikit-learn/pull/24100#discussion_r966045459. When `max_samples` is an int, we do not accept 0, so I do not think we should support a fractional `max_samples` that rounds down to a zero either.

This means we can check `max_samples` earlier in `_get_n_samples_bootstrap`. This is the solution proposed in #25140.